### PR TITLE
Remove all wrong usage of singlr flight; also fix voting power cache key

### PR DESCRIPTION
**Pull request by rlan35**
Single flight is meant for duplicate func call avoidance (same-time duplicate calls). it's not cache. The func call results are not cached (https://medium.com/@vCabbage/go-avoid-duplicate-requests-with-sync-singleflight-311601b3068b)

Found the issue after seeing lots of accumulated pending goroutines in our node. They are caused by the sleep() and forget() part of singleFlight. Since it's not cache, actually the func got called many times: 
image

Besides the rpc layer, all other part of the singleFlight usage are wrong. Should instead just use lru cache.

This PR also fix the cache key on voting power cache which was hidden due to that single flight didn't do the cache work for voting power.

Tested locally.

### CODE REVIEW

File: internal/chain/reward.go [hunk 1]
```diff
@@ -45,34 +46,35 @@ func ballotResultBeaconchain(
...
-			if err != nil {
-				return nil, err
-			}
-
-			// For new calc, remove old data from 3 epochs ago
-			deleteEpoch := big.NewInt(0).Sub(epoch, big.NewInt(3))
-			deleteKey := fmt.Sprintf("%s-%d", deleteEpoch.String(), subComm.ShardID)
-			votingPowerCache.Forget(deleteKey)
+	// Look up
+	key := fmt.Sprintf("%s-%d", epoch.String(), subComm.Hash())
```
Comment (JackyWYX) [inline internal/chain/reward.go:57]: format `"%s-%x"`?

File: internal/chain/reward.go [hunk 2]
```diff
@@ -45,34 +46,35 @@ func ballotResultBeaconchain(
...
+	fmt.Println(key)
+	if b, ok := votingPowerCache.Get(key); ok {
+		if b == nil {
+			fmt.Println("DAMN")
```
Comment (LeoHChen) [inline internal/chain/reward.go:61]: log this error please

File: internal/chain/reward.go [hunk 3]
```diff
@@ -45,34 +46,35 @@ func ballotResultBeaconchain(
 }
 
 var (
-	votingPowerCache   singleflight.Group
-	delegateShareCache singleflight.Group
+	votingPowerCache, _   = lru.New(16)
```
Comment (LeoHChen) [inline internal/chain/reward.go:49]: why 16 here?  can you elaborate it a bit?

File: internal/chain/reward.go [hunk 4]
```diff
@@ -45,34 +46,35 @@ func ballotResultBeaconchain(
...
+	votingPower, err := votepower.Compute(subComm, epoch)
 	if err != nil {
 		return nil, err
 	}
 
-	return results.(*votepower.Roster), nil
+	// Put in cache
+	votingPowerCache.Add(key, votingPower)
+	if votingPower == nil {
+		fmt.Println("DAMN2")
```
Comment (LeoHChen) [inline internal/chain/reward.go:75]: log the error please

File: internal/chain/reward.go [hunk 5]
```diff
@@ -81,40 +83,35 @@ func lookupDelegatorShares(
...
+	}
 
-			for i := range validatorSnapshot.Delegations {
-				delegation := validatorSnapshot.Delegations[i]
-				// NOTE percentage = <this_delegator_amount>/<total_delegation>
-				percentage := numeric.NewDecFromBigInt(delegation.Amount).Quo(totalDelegationDec)
-				result[delegation.DelegatorAddress] = percentage
-			}
+	// If not found, construct
+	result := map[common.Address]numeric.Dec{}
```
Comment (LeoHChen) [inline internal/chain/reward.go:95]: will the map cause concurrency issue?

Comment (LeoHChen) [reply internal/chain/reward.go:95]: do you need to protect it?

File: internal/chain/reward.go (hunk 2 above)
Comment (rlan35) [reply internal/chain/reward.go]: sorry, those are debug code. Forgot to remove. Removed.

File: internal/chain/reward.go (hunk 1 above)
Comment (rlan35) [reply internal/chain/reward.go]: done.

File: internal/chain/reward.go (hunk 3 above)
Comment (rlan35) [reply internal/chain/reward.go:51]: votingPower is the calcuation per committee (or per epoch/shard). so 16 is good enough for multiple epochs.

File: internal/chain/reward.go (hunk 5 above)
Comment (rlan35) [reply internal/chain/reward.go:90]: this map is atomically created and only read, never updated. So no concurrency issue.

**Comment by rlan35** [comment]
tested on a node connected to mainnet shard 0. The goroutine usage all looks normal now.

**Pull request merged**